### PR TITLE
New: Wrap custom-style

### LINKED ISSUE #155: Wrap <style is="custom-style"> in <custom-style>
A hybrid rule / upgrade pass.

### PR DESCRIPTION
**Pull request by stramel**
- [x] CHANGELOG.md has been updated

Fixes #155 
Relates to #142 

/cc @rictic 

Not sure which collections this should be in

---
This change is [Reviewable](https://reviewable.io/reviews/polymer/polymer-linter/166)

### CODE REVIEW

File: src/test/polymer/custom-style-extension_test.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  test('warns for the proper cases and with the right messages', async() => {
+    const {warnings} = await linter.lint([`${ruleId}/${ruleId}.html`]);
+    assert.deepEqual(warningPrinter.prettyPrint(warnings), [
+      `
+<style is="custom-style">
+       ~~~~~~~~~~~~~~~~~`,
+    ]);
+
+    assert.deepEqual(warnings.map((w) => w.message), [
+      '<style> should not be extended with \`is="custom-style"\` but instead wrapped with \`<custom-style>\`.',
```
Comment (rictic) [inline src/test/polymer/custom-style-extension_test.ts]: Hm, for hybrid code I think we might want both so let's not discourage `is="custom-style"` here, just encourage wrapping with `<custom-style>`

File: src/polymer/custom-style-extension.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
```
Comment (rictic) [inline src/polymer/custom-style-extension.ts]: We're switching over to the new iteration-based dom5 API. 

`import * as dom5 from 'dom5/lib/index-next';`

File: src/polymer/custom-style-extension.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                --app-color: red;
+              }
+            </style>
+          </custom-style>
+  `);
+
+  async checkDocument(parsedDocument: ParsedHtmlDocument, _document: Document) {
+    const warnings: Warning[] = [];
+
+    const customStyleTags = dom5.queryAll(
```
Comment (rictic) [inline src/polymer/custom-style-extension.ts]: New dom5 api returns an iterable, but we can get an array easily enough out of it:

```typescript
const customStyleTags = [...dom5.queryAll(parsedDocument.ast, isCustomStyleV1, dom5.childNodesIncludeTemplate)];
```

File: test/custom-style-extension/before-fixes.html [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<style is="custom-style">
```
Comment (rictic) [inline test/custom-style-extension/before-fixes.html:1]: there's some tricky indentation code in this PR. let's add in another style node that already has some indentation, to ensure it's handling it correctly

Comment (stramel) [reply test/custom-style-extension/before-fixes.html:1]: Good call! Wasn't properly handling that case and realized I left some duplicate code in there.

**Comment by stramel** [comment]
Build seems to be failing on the difference between `\r\n` and `\n`

File: src/polymer/custom-style-extension.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      const fix: Replacement[] =
+          indentSourceRange(sourceRange, `  `, parsedDocument);
+
+      fix.push({
+        range: {
+          file: parsedDocument.url,
+          start: sourceRange.start,
+          end: sourceRange.start
+        },
+        replacementText: `<custom-style>\r\n${indentation}`
```
Comment (rictic) [inline src/polymer/custom-style-extension.ts]: Here and elsewhere, `\r\n` -> `\n`

If we want to handle windows newlines we should do it in the code that applies edits (detect line endings the file uses, replace `\n` with `\r\n` in replacement text when building up the new version of the file)

Comment (stramel) [reply src/polymer/custom-style-extension.ts]: Doh, shoot! I forgot I changed that locally. Good catch!

**Comment by stramel** [comment]
For CLA purposes, I'm alright with @rictic's changes being in my PR.

**Comment by rictic** [comment]
Great!

**Review by rictic — APPROVED** [review]

**Pull request merged**